File: python/vision_forge/services/nodes/frameworks/tensorflow/nodes/conv2d.py
"""TensorFlow Conv2D Layer Node"""

from typing import Any, Dict, List, Optional

from ....core.types import ConfigField, ConfigOption, Framework, NodeMetadata, TensorShape
from ..base import TensorFlowNodeDefinition


class Conv2DNode(TensorFlowNodeDefinition):
    """
    2D Convolutional Layer.

    Uses tf.keras.layers.Conv2D.
    TensorFlow uses NHWC format by default.

    Input: [batch_size, height, width, channels]
    Output: [batch_size, new_height, new_width, filters]
    """

    @property
    def metadata(self) -> NodeMetadata:
        return NodeMetadata(
            type="conv2d",
            label="Conv2D",
            category="basic",
            color="var(--color-purple)",
            icon="SquareHalf",
            description="2D convolutional layer",
            framework=Framework.TENSORFLOW
        )

    @property
    def config_schema(self) -> List[ConfigField]:
        return [
            ConfigField(
                name="filters",
                label="Filters",
                type="number",
                required=True,
                min=1,
                description="Number of output filters"
            ),
            ConfigField(
                name="kernel_size",
                label="Kernel Size",
                type="number",
                default=3,
                min=1,
                description="Size of convolving kernel"
            ),
            ConfigField(
                name="strides",
                label="Strides",
                type="number",
                default=1,
                min=1,
                description="Stride of convolution"
            ),
            ConfigField(
                name="padding",
                label="Padding",
                type="select",
                default="valid",
                options=[
                    ConfigOption(value="valid", label="Valid"),
                    ConfigOption(value="same", label="Same")
                ],
                description="Padding mode"
            ),
            ConfigField(
                name="activation",
                label="Activation",
                type="select",
                default="None",
                options=[
                    ConfigOption(value="None", label="None"),
                    ConfigOption(value="relu", label="ReLU"),
                    ConfigOption(value="sigmoid", label="Sigmoid"),
                    ConfigOption(value="tanh", label="Tanh")
                ],
                description="Activation function"
            )
        ]

    def compute_output_shape(
        self,
        input_shape: Optional[TensorShape],
        config: Dict[str, Any]
    ) -> Optional[TensorShape]:
        filters = config.get("filters")
        if not input_shape or filters is None or filters == '':
            return None

        # TensorFlow uses NHWC format
        if len(input_shape.dims) != 4:
            return None

        try:
            num_filters = int(filters)
            if num_filters <= 0:
                return None
        except (ValueError, TypeError):
            return None

        batch, height, width, _ = input_shape.dims
        kernel = int(config.get("kernel_size", 3))
        strides = int(config.get("strides", 1))
        padding = config.get("padding", "valid")

        # Calculate output dimensions
        if padding == "same":
            out_height = (height + strides - 1) // strides
            out_width = (width + strides - 1) // strides
        else:  # valid
            out_height = (height - kernel) // strides + 1
            out_width = (width - kernel) // strides + 1

        return TensorShape(
            dims=[batch, out_height, out_width, num_filters],
            description="Convolved feature map (NHWC)"
        )

    def validate_incoming_connection(
        self,
        source_node_type: str,
        source_output_shape: Optional[TensorShape],
        target_config: Dict[str, Any]
    ) -> Optional[str]:
        if source_node_type in ("input", "dataloader", "empty", "custom"):
            return None

        return self.validate_dimensions(
            source_output_shape,
            4,
            "[batch, height, width, channels]"
        )

File: packages/core/src/lib/nodes/definitions/pytorch/conv2d.ts
/**
 * PyTorch Conv2D Layer Node Definition
 * Enhanced with pattern-based validation
 */

import { NodeDefinition } from '../../base'
import { NodeMetadata, BackendFramework } from '../../contracts'
import { TensorShape, BlockConfig, ConfigField, BlockType, ShapePattern, DimensionValue } from '../../../types'
import { spatialInput } from '../../../validation/patterns'
import { getRank, isNumeric } from '../../../validation/matchers'

export class Conv2DNode extends NodeDefinition {
  readonly metadata: NodeMetadata = {
    type: 'conv2d',
    label: 'Conv2D',
    category: 'basic',
    color: 'var(--color-purple)',
    icon: 'SquareHalf',
    description: '2D convolutional layer',
    framework: BackendFramework.PyTorch
  }

  /**
   * Input pattern: 4D spatial tensor (B, C, H, W)
   */
  readonly inputPattern: ShapePattern = spatialInput()

  readonly configSchema: ConfigField[] = [
    {
      name: 'in_channels',
      label: 'Input Channels',
      type: 'number',
      required: false,
      min: 1,
      description: 'Number of input channels (auto-inferred if not set)'
    },
    {
      name: 'out_channels',
      label: 'Output Channels',
      type: 'number',
      required: true,
      min: 1,
      description: 'Number of output channels'
    },
    {
      name: 'kernel_size',
      label: 'Kernel Size',
      type: 'number',
      default: 3,
      min: 1,
      description: 'Size of convolving kernel'
    },
    {
      name: 'stride',
      label: 'Stride',
      type: 'number',
      default: 1,
      min: 1,
      description: 'Stride of convolution'
    },
    {
      name: 'padding',
      label: 'Padding',
      type: 'number',
      default: 0,
      min: 0,
      description: 'Zero-padding added to both sides'
    },
    {
      name: 'dilation',
      label: 'Dilation',
      type: 'number',
      default: 1,
      min: 1,
      description: 'Spacing between kernel elements'
    }
  ]

  computeOutputShape(inputShape: TensorShape | undefined, config: BlockConfig): TensorShape | undefined {
    if (!inputShape || !config.out_channels) {
      return undefined
    }

    const rank = getRank(inputShape)
    if (rank !== 4) {
      return undefined
    }

    const [batch, , height, width] = inputShape.dims
    const kernel = (config.kernel_size as number) || 3
    const stride = (config.stride as number) || 1
    const padding = (config.padding as number) || 0
    const dilation = (config.dilation as number) || 1
    const outChannels = config.out_channels as number

    // Compute output dimensions
    let outHeight: DimensionValue
    let outWidth: DimensionValue

    if (isNumeric(height)) {
      const [h] = this.computeConv2DOutput(height, 1, kernel, stride, padding, dilation)
      outHeight = h
    } else {
      outHeight = `${height}'`
    }

    if (isNumeric(width)) {
      const [, w] = this.computeConv2DOutput(1, width, kernel, stride, padding, dilation)
      outWidth = w
    } else {
      outWidth = `${width}'`
    }

    return {
      dims: [batch, outChannels, outHeight, outWidth],
      description: 'Convolved feature map',
      flags: {
        inferred: true,
        symbolic: !isNumeric(outHeight) || !isNumeric(outWidth)
      },
      provenance: {
        source: 'computed',
        transformation: 'conv2d'
      }
    }
  }

  validateIncomingConnection(
    sourceNodeType: BlockType,
    sourceOutputShape: TensorShape | undefined,
    targetConfig: BlockConfig
  ): string | undefined {
    // Allow connections from input/dataloader without shape validation
    if (sourceNodeType === 'input' || sourceNodeType === 'dataloader') {
      return undefined
    }

    // Empty and custom nodes are flexible
    if (sourceNodeType === 'empty' || sourceNodeType === 'custom') {
      return undefined
    }

    if (!sourceOutputShape) {
      return undefined
    }

    const rank = getRank(sourceOutputShape)
    if (rank !== 4) {
      return `Conv2D requires 4D input (B, C, H, W), got ${rank}D`
    }

    // Check input channels if specified
    const inChannels = targetConfig.in_channels as number | undefined
    if (inChannels && inChannels > 0) {
      const inputChannels = sourceOutputShape.dims[1]
      if (isNumeric(inputChannels) && inputChannels !== inChannels) {
        return `Channel mismatch: input has ${inputChannels} channels, expected ${inChannels}`
      }
    }

    return undefined
  }

  getDefaultConfig(): BlockConfig {
    return {
      out_channels: '',
      kernel_size: 3,
      stride: 1,
      padding: 0,
      dilation: 1
    }
  }
}
